### PR TITLE
Add new module macrel/contigs

### LINKED ISSUE #929: new module: macrel
I think it would be good to have a module for [macrel](https://bioconda.github.io/recipes/macrel/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [ ] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Darcy220606**
## PR checklist

Closes #929 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: modules/macrel/contigs/meta.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      description: A fasta file cntaiing aminoacid sequences showing the general gene prediction information in the contigs.
+      pattern: "smorfs.faa"
+  - all_orfs:
+      type: file
+      description: A fasta file cntaiing aminoacid sequences showing the general gene prediction information in the contigs.
+      pattern: "all_orfs.faa"
+  - readme_file:
+      type: file
+      description: A readme file containing tool specific information (e.g. citations, details about the output, etc.).
+      pattern: "*.{md}"
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:54]:
```suggestion
      pattern: "*.md"
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
- Does `  --keep-negatives      Whether to keep non-AMPs in the output` result in extra files?
- Why is a bunch of things deleted from `tests/config/pytest_modules.yml`? I would keep that all the same as in dev except for inserting the macrel lines
- You need to run your `nf-core modules create-test-yml macrel` command to update the test results ,current it's left as ` bam` file which I'm assuming is incorrect? I'm not exactly sure how/why your tests have passed in this case :thinking:

**Comment by Darcy220606** [comment]
> 

Thx @jfy133  ! Regarding your comments:
- --keep-negatives : actually writes non AMPs in the same predictions file
-  I checked it again and it is the same as the pytest_modules.yml from the master branch -with those hashes after some tools and unsorted tools. Is their another draft i should use instead? 
- I updated the test.yml. 
- --mem is for memory allocation, should i also pass it ${task.cpus}? I thought that was only for threading?

**Comment by jfy133** [comment]
> --keep-negatives : actually writes non AMPs in the same predictions file

Awesome! Then no changed neeed there.

> I checked it again and it is the same as the pytest_modules.yml from the master branch -with those hashes after some tools and unsorted tools. Is their another draft i should use instead?

See slack, your diff suggests otherwise.

> I updated the test.yml.

OK, tests still fail. Will need to look why.

> --mem is for memory allocation, should i also pass it ${task.cpus}? I thought that was only for threading?

Oops, shouldn't review from my phone :grimacing: I meant `${task.memory}`

File: tests/modules/macrel/contigs/test.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+      md5sum: fa3706dfc95d0538a52c4d0d824be5fb
+    - path: output/macrel/test/test.all_orfs.faa
+      md5sum: 1dafaa08e76532ea5e94ec742c7ca5e5
+    - path: output/macrel/test/test.prediction.gz
+    - path: output/macrel/test/test.smorfs.faa
+      md5sum: 59aacca46e94ffca8d765da5e28afc28
+    - path: output/macrel/test/test_log.txt
+      md5sum: 6fdba143dce759597eb9f80e5d968729
+    - path: output/macrel/versions.yml
+      md5sum: 6b9071d9574f002cc906dbb9ce125eed
```
Comment (jfy133) [inline tests/modules/macrel/contigs/test.yml:17]:
```suggestion
      md5sum: be8bf0d0647751c635c3736655f29f85
```

File: modules/macrel/contigs/main.nf [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"
+    """
+    macrel contigs \\
+        $args \\
+        --fasta $fasta \\
+        --output ${prefix}/ \\
+        --tag ${prefix} \\
+        --log-file ${prefix}/${prefix}_log.txt \\
```
Comment (jfy133) [inline modules/macrel/contigs/main.nf:30]:
```suggestion
        --log-file ${prefix}/${prefix}_log.txt \\
        --mem ${task.memory}
```

File: modules/macrel/contigs/main.nf [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        --mem ${task.memory}
```
Comment (jfy133) [inline modules/macrel/contigs/main.nf:31]:
```suggestion
```

**Comment by jfy133** [comment]
Ok ignore the `mem` thing, 
```
--mem: Used during assembly to specify the percent of RAM used (1-0)
```

Found that on their hidden readthedocs. Not really sure what that means so safer to leave it ;)

**Comment by Darcy220606** [comment]
Thats great - it now works Thx a bunch @jfy133 . It was the last hash in the test.yml all along :)

File: modules/macrel/contigs/main.nf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/macrel:1.1.0--py36hc5360cc_0':
+        'quay.io/biocontainers/macrel:1.1.0--py36hc5360cc_0' }"
+
+    input:
+    tuple val(meta), path(fasta)
+
+    output:
+    tuple val(meta), path("*/*.smorfs.faa")     , emit: smorfs
+    tuple val(meta), path("*/*.all_orfs.faa")   , emit: all_orfs
```
Comment (jfy133) [inline modules/macrel/contigs/main.nf:15]: Would be good to compress everything in `.faa` after the main macrel command has finished :)

File: modules/macrel/contigs/main.nf [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        $args \\
+        --fasta $fasta \\
+        --output ${prefix}/ \\
+        --tag ${prefix} \\
+        --log-file ${prefix}/${prefix}_log.txt \\
+        --threads $task.cpus
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        macrel: \$(echo \$(macrel --version | sed 's/macrel//g'))
```
Comment (jfy133) [inline modules/macrel/contigs/main.nf:35]:
```suggestion
        macrel: \$(echo \$(macrel --version | sed 's/macrel //g'))
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
Just need to compress the output and commit the one typo and we are good to go :D

File: modules/macrel/contigs/main.nf (hunk 5 above)
Comment (Darcy220606) [reply modules/macrel/contigs/main.nf]: Ok- all done now :)

File: tests/modules/macrel/contigs/test.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+- name: macrel contigs test_macrel_contigs
+  command: nextflow run ./tests/modules/macrel/contigs -entry test_macrel_contigs -c ./tests/config/nextflow.config -c ./tests/modules/macrel/contigs/nextflow.config
+  tags:
+    - macrel/contigs
+    - macrel
+  files:
+    - path: output/macrel/test/README.md
+      md5sum: fa3706dfc95d0538a52c4d0d824be5fb
+    - path: output/macrel/test/test.all_orfs.faa.gz
+      md5sum: 644299f5b89125e6bde42137c76249c1
```
Comment (jfy133) [inline tests/modules/macrel/contigs/test.yml:10]:
```suggestion
```

File: modules/macrel/contigs/main.nf [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    tuple val(meta), path("*/*.smorfs.faa.gz")      , emit: smorfs
+    tuple val(meta), path("*/*.all_orfs.faa.gz")    , emit: all_orfs
+    tuple val(meta), path("*/*.prediction.gz")      , emit: amp_prediction
+    tuple val(meta), path("*/*.md")                 , emit: readme_file
+    tuple val(meta), path("*/*_log.txt")            , emit: log_file
+    path "versions.yml"                             , emit: versions
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"
```
Comment (jfy133) [inline modules/macrel/contigs/main.nf:23]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta.id}"
```

File: modules/macrel/contigs/meta.yml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - versions:
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - amp_prediction:
+      type: file
+      description: A zipped file, with all predicted amps in a table format.
+      pattern: "prediction.gz"
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:42]:
```suggestion
      pattern: "*.prediction.gz"
```

File: modules/macrel/contigs/meta.yml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  - smorfs:
+      type: file
+      description: A zipped fasta file containing aminoacid sequences showing the general gene prediction information in the contigs.
+      pattern: "smorfs.faa"
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:46]:
```suggestion
      pattern: "*.smorfs.faa.gz"
```

File: modules/macrel/contigs/meta.yml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+  - all_orfs:
+      type: file
+      description: A zipped fasta file containing aminoacid sequences showing the general gene prediction information in the contigs.
+      pattern: "all_orfs.faa"
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:50]:
```suggestion
      pattern: "*.all_orfs.faa.gz"
```

File: modules/macrel/contigs/meta.yml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      type: file
+      description: A zipped file, with all predicted amps in a table format.
+      pattern: "prediction.gz"
+  - smorfs:
+      type: file
+      description: A zipped fasta file containing aminoacid sequences showing the general gene prediction information in the contigs.
+      pattern: "smorfs.faa"
+  - all_orfs:
+      type: file
+      description: A zipped fasta file containing aminoacid sequences showing the general gene prediction information in the contigs.
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:49]:
```suggestion
      description: A zipped fasta file containing amino acid sequences showing the general gene prediction information in the contigs.
```

File: modules/macrel/contigs/meta.yml [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+      pattern: "all_orfs.faa"
+  - readme_file:
+      type: file
+      description: A readme file containing tool specific information (e.g. citations, details about the output, etc.).
+      pattern: "*.md"
+  - log_file:
+      type: file
+      description: A log file containing the information pertainign to the run.
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:57]:
```suggestion
      description: A log file containing the information pertaining to the run.
```

File: modules/macrel/contigs/meta.yml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      pattern: "log.txt"
```
Comment (jfy133) [inline modules/macrel/contigs/meta.yml:58]:
```suggestion
      pattern: "*_log.txt"
```

**Review by jfy133 — APPROVED** [review]
Please commit the final metadata changes and you're good to go!

**Pull request merged**